File: app/src/main/java/com/example/moxfeild/MainActivity.java
package com.example.moxfeild;  // Make sure this matches your actual package name

import android.os.Bundle;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import androidx.activity.ComponentActivity;

public class MainActivity extends ComponentActivity {
    private WebView webView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Create WebView
        webView = new WebView(this);
        setContentView(webView);

        // Configure WebView with a custom WebViewClient that suppresses errors
        webView.setWebViewClient(new WebViewClient() {
            // This method prevents error pages from being displayed
            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                // Do nothing - this suppresses the error display

                // Optional: Silently retry loading after a brief delay
                if (request.isForMainFrame()) {
                    view.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            webView.reload();
                        }
                    }, 5000); // Try again after 5 seconds
                }
            }
        });

        // Configure WebView settings
        WebSettings webSettings = webView.getSettings();
        webSettings.setJavaScriptEnabled(true); // Enable JavaScript for full functionality
        webSettings.setDomStorageEnabled(true); // Enable DOM storage for modern web apps
        webSettings.setLoadWithOverviewMode(true); // Load pages to fit in the screen
        webSettings.setUseWideViewPort(true); // Use wide viewport for better rendering
        webSettings.setBuiltInZoomControls(true); // Enable zoom controls
        webSettings.setDisplayZoomControls(false); // Hide zoom controls after a moment

        // Optional: Cache the website for offline use
        webSettings.setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);

        // Load the Moxfield website
        webView.loadUrl("https://www.moxfield.com/decks/personal");
    }

    // Override the back button to navigate within the WebView
    @Override
    @Deprecated
    public void onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack(); // Go back in WebView if possible
        } else {
            super.onBackPressed(); // Otherwise, exit the app
        }
    }
}

File: app/src/main/java/com/example/moxfeild/MainActivity.kt
package com.example.moxfeild

import android.os.Bundle
import android.webkit.CookieManager
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.ComponentActivity
import java.io.ByteArrayInputStream

class MainActivity : ComponentActivity() {
    private lateinit var webView: WebView

    private val adDomains = listOf(
        "doubleclick.net",
        "googleadservices.com",
        "googlesyndication.com",
        "adservice.google.com",
        "google-analytics.com",
        "googletagmanager.com",
        "adform.net",
        "adnxs.com",
        "criteo.com",
        "amazon-adsystem.com",
        "taboola.com",
        "outbrain.com"
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        webView = WebView(this)
        setContentView(webView)

        // Configure CookieManager before setting up WebView
        val cookieManager = CookieManager.getInstance()
        cookieManager.setAcceptCookie(true)
        cookieManager.setAcceptThirdPartyCookies(webView, true)

        webView.webViewClient = object : WebViewClient() {
            override fun shouldInterceptRequest(view: WebView?, request: WebResourceRequest?): WebResourceResponse? {
                val url = request?.url?.toString() ?: ""
                if (adDomains.any { url.contains(it) }) {
                    return WebResourceResponse(
                        "text/plain",
                        "utf-8",
                        ByteArrayInputStream("".toByteArray())
                    )
                }
                return super.shouldInterceptRequest(view, request)
            }

            override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
                if (request?.isForMainFrame == true) {
                    view?.postDelayed({
                        webView.reload()
                    }, 5000)
                }
            }

            // Add page finished handler to ensure cookies are saved
            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                cookieManager.flush() // Ensure cookies are persisted to storage
            }
        }

        val webSettings = webView.settings
        webSettings.javaScriptEnabled = true
        webSettings.domStorageEnabled = true
        webSettings.loadWithOverviewMode = true
        webSettings.useWideViewPort = true
        webSettings.builtInZoomControls = true
        webSettings.displayZoomControls = false
        webSettings.setSupportMultipleWindows(true)
        webSettings.databaseEnabled = true

        // For modern Android: use default cache mode
        webSettings.cacheMode = WebSettings.LOAD_DEFAULT

        // Add these for better session persistence
        webSettings.allowFileAccess = true
        webSettings.saveFormData = true

        if (savedInstanceState != null) {
            webView.restoreState(savedInstanceState)
        } else {
            webView.loadUrl("https://moxfield.com/decks/personal")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        webView.saveState(outState)
    }

    override fun onPause() {
        super.onPause()
        CookieManager.getInstance().flush() // Ensure cookies are saved when app is paused
    }

    override fun onResume() {
        super.onResume()
        // Optionally refresh the page if needed
        // webView.reload()
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack()
        } else {
            super.onBackPressed()
        }
    }
}